src/ucs/datastruct/usage_tracker.c

```diff
@@ -0,0 +1,244 @@
+}
+
+/* Insert an entry to the hash table and update its score. */
+static ucs_usage_tracker_element_t *
+ucs_usage_tracker_put(ucs_usage_tracker_h usage_tracker, void *key)
+{
+    int khret;
+    khiter_t iter;
+    ucs_usage_tracker_element_t *elem;
+
+    iter = kh_put(usage_tracker_hash, &usage_tracker->hash, (uint64_t)key,
+                  &khret);
+    ucs_assert(khret != UCS_KH_PUT_FAILED);
+
+    elem = &kh_val(&usage_tracker->hash, iter);
+
+    if ((khret == UCS_KH_PUT_BUCKET_EMPTY) ||
+        (khret == UCS_KH_PUT_BUCKET_CLEAR)) {
+        elem->score     = usage_tracker->params.exp_decay.b;
+        elem->min_score = 0;
+        elem->promoted  = 0;
+        elem->key       = key;
+    }
+
+    return elem;
+}
+
+ucs_status_t ucs_usage_tracker_get_score(ucs_usage_tracker_h usage_tracker,
+                                         void *key, double *score_p)
+{
+    ucs_usage_tracker_element_t *item;
+    khiter_t iter;
+
+    iter = kh_get(usage_tracker_hash, &usage_tracker->hash, (uint64_t)key);
+    if (iter == kh_end(&usage_tracker->hash)) {
+        return UCS_ERR_NO_ELEM;
+    }
+
+    item     = &kh_value(&usage_tracker->hash, iter);
+    *score_p = ucs_usage_tracker_score(item);
+    return UCS_OK;
+}
+
+ucs_status_t
+ucs_usage_tracker_remove(ucs_usage_tracker_h usage_tracker, void *key)
+{
+    khiter_t iter;
+
+    iter = kh_get(usage_tracker_hash, &usage_tracker->hash, (uint64_t)key);
+    if (iter == kh_end(&usage_tracker->hash)) {
+        return UCS_ERR_NO_ELEM;
+    }
+
+    kh_del(usage_tracker_hash, &usage_tracker->hash, iter);
+    return UCS_OK;
+}
+
+void ucs_usage_tracker_push_min_score(ucs_usage_tracker_h usage_tracker,
+                                      void *key, double score)
+{
+    ucs_usage_tracker_element_t *element;
+
+    element            = ucs_usage_tracker_put(usage_tracker, key);
+    element->min_score = score;
+}
+
+/* Checks if an entry has high enough score to get promoted. */
+static int
+ucs_usage_tracker_compare(const void *elem1, const void *elem2, void *arg)
+{
+    const ucs_usage_tracker_params_t *params = arg;
+    ucs_usage_tracker_element_t *item1 = *(ucs_usage_tracker_element_t**)elem1;
+    ucs_usage_tracker_element_t *item2 = *(ucs_usage_tracker_element_t**)elem2;
+
+    if (fabs(item1->score - item2->score) < params->remove_thresh) {
+        return (item1->key > item2->key) ? 1 : -1;
+    }
+
+    return (item1->score < item2->score) ? 1 : -1;
+}
+
+/* Promote/Demote entires base on the latest score, and triggers user
+  * callback accordingly. */
+static void ucs_usage_tracker_update_rank(ucs_usage_tracker_h usage_tracker)
+{
+    const ucs_usage_tracker_params_t *params = &usage_tracker->params;
+    size_t elem_index                        = 0;
+    ucs_usage_tracker_element_t **elems_array, *item;
+    khiter_t iter;
+    uint64_t key;
+
+    elems_array = ucs_alloca(sizeof(*elems_array) *
+                             (kh_size(&usage_tracker->hash)));

```

Comment:
1. extra ( )
2. better use malloc, this is not fast path. also, we can use ucs_array_t